File: algorithms/[179]最大数/solution.ts
// topics = ["数组"]

function largestNumber(nums: number[]): string {
  nums.sort((a, b) => {
    const sa = String(a);
    const sb = String(b);
    if (sa + sb > sb + sa) {
      return -1;
    }
    if (sa + sb < sb + sa) {
      return 1;
    }
    return 0;
  });

  let str = nums.join('');

  // 排除类似 '00' 这种情况
  while (str.startsWith('0') && str.length > 1) {
    str = str.slice(1);
  }

  return str;
}

File: algorithms/[179]最大数/solution.py
# topics = ["数组"]

from functools import cmp_to_key
from typing import List


class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        def compare(a: int, b: int) -> int:
            x = y = 10
            while x <= a:
                x *= 10
            while y <= b:
                y *= 10

            return x * b + a - (y * a + b)

        nums.sort(key=cmp_to_key(compare))

        if nums[0] == 0:
            return '0'

        return ''.join(map(str, nums))
